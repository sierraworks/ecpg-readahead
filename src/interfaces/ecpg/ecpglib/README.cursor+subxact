Theory of operation for client side cursor accounting and readahead
===================================================================

Contents of tuples returned by a cursor always reflect the data present at
the time of declaring the cursor. Because of this, executing DML commands
involving the table(s) appearing in the cursor's SELECT query won't modify
the result set returned by the cursor. Because of this, cursor readahead is
possible and the cursor readahead window always contains the same data as
the underlying cursor.

Tracking the maximum known backend position value allows the number of MOVE
commands sent to the server to be reduced and re-fetching the same readahead
window avoided in case the application suddenly changes cursor positioning
from positive to negative and vice-versa.

The idea behind cursor readahead is to move one tuple less than asked by
the application and then ecpglib performs a FETCH FORWARD/BACKWARD N to
populate the local tuple cache.

By using a readahead window larger than 1 tuple, the cursor positions known
by the backend and the application are different most of the time. As a
consequence, UPDATE/DELETE ... WHERE CURRENT OF requires a MOVE behind the
application's back. This may reduce or eliminate the caching effects in the
cursor and it will not work for non-scrollable cursors. Either the cursor
must be scrollable or READAHEAD 1 must be explicitly specified for such
cursors.

The slight problem with this approach is that ABSOLUTE and RELATIVE movements
needs two operations. If the next command issued by the application spills
out of the cache and it happens frequently, it may actually decrease
performance. To prevent this, the ECPG runtime switches to executing these
kinds of commands as is after 3 cache misses. FETCH FORWARD/BACKWARD allows
switching back to caching.

The ECPG runtime library has "struct subxact_descriptor" and
"struct cursor_descriptor" for tracking subtransactions and cursors.

Explanation of structure members of "struct cursor_descriptor" are below.

subxact_level: The cursor was declared in this transaction level.
--------------
Valid values are:
0:  Outside of transactions. This is a WITH HOLD cursor, committed
    in a previous transaction.
1:  The cursor was declared in the currently running toplevel
    transaction.
>2: The cursor was declared in the currently running transaction
    in the specified SAVEPOINT.

As soon as a (sub-)transaction is committed or rolled back, cursors are
either implicitly propagated to the next subtransaction level upwards,
i.e. subxact_level is decreased by 1.

with_hold: WITH HOLD cursor (true or false)
----------

scrollable: The scrollability of the cursor.
-----------
Accepted values are: ECPGcs_no_scroll or ECPGcs_scroll

There is also a third enum value, ECPGcs_unspecified which is used
internally by the ECPG preprocessor when the application didn't
explicitly specify neither SCROLL nor NO SCROLL. It will be eventually
overridden by NO SCROLL in the DECLARE statement and ECPGcs_no_scroll
in the "scrollable" argument.

res: The current readahead window.
----
It is used to return data across several FETCH commands, speeding up
the application by reducing network turnaround.

readahead: The size of the readahead window.
----------
The ECPG runtime library attempts to fetch this amount of records at once.

cache_start_pos: The start position of the readahead window.
----------------
The value is 1-based. Position 0 means that the cursor is before the first
tuple in the result set.

I can also be negative (but also 1-based) if the application started with
FETCH/MOVE ABSOLUTE <some negative number>. In this case the numbering
is used to count from the end of the result set.

backward: The ECPG runtime library fetched this readahead window backward.
---------

atstart: The current position is at the start of the result set.
--------

atend: The current position is at the end of the result set.
------

pos: The current cursor position, as known by the application.
----
The comments above for "cache_start_pos" are also applicable here.
"atstart" and "atend" apply to this position.

backend_pos: The current cursor position, as known by the server.
------------

mkbpos: The maximum known backend position.
-------
If the application tries to execute a MOVE and it's over the previously
known maximum value, different things happen depending on the type of the
motion.

MOVE [FORWARD|BACKWARD] ALL
The number of records from the previous position to the end of the result
set is reported.

MOVE ABSOLUTE N, MOVE [FORWARD|BACKWARD] RELATIVE N
If there is a tuple on this position, the server indicates the existence of
a tuple on that position.

This value is trivially updated for FETCH commands.

mkbp_is_last: Maximum Known Backend Position is the last valid position.
-------------
The "mkbpos" is the last cursor position where a tuple exists 

Structure members of "struct subxact_descriptor" don't need too much
explanation. New savepoints get prepended to the list of subtransations
which are per-connection, so the head of the list is found in
struct connection. However, the necessity of this structure is explained
below.

Tracking (sub-)transactions
===========================

One side effect/requirement of tracking cursors in the ECPG runtime library
is that (sub-)transactions are also needed to be tracked. These two are
in some kind of symbiosis: if a command fails, the (sub-)transaction returns
"current transaction is aborted" error for every subsequent commands.
Such a failed command can be a cursor command using an invalid cursor name.

By tracking cursors in the ECPG runtime library, detecting an invalid cursor
name can be done early which reduces the network turnaround further.

But doing this, the server and the application now disagree on the failed
transaction. Because of this, a "client side error" flag was introduced into
"struct connection".

Subsequent commands (in execute.c::ecpg_prologue() and misc.c::ECPGtrans())
check this flag and return the "current transaction is aborted" error to the
application without executing them in the server. As mentioned above, the
server and the application at this point disagree on the failed transaction
and he issued command may be executed without an error, causing unwanted
effects.

Checking the "client side error" flag further reduces network turnaround for
transactions failed on the client side and buggy applications that don't check
for failed commands and roll back the (sub-)transaction.

Executing COMMIT or PREPARE TRANSACTION for a failed transaction silently
overrides these commands as ROLLBACK in the server. By checking the "client
side error" flag, it can be done explicitly.

In a failed transaction, only the ROLLBACK and ROLLBACK TO SAVEPOINT commands
are allowed to be executed without errors in the server. These also pass
through in client side tracking code, allowing the "client side error" flag
to be cleared.

All other transaction commands are treated as regular commands in a failed
transactions in that the "current transaction is aborted" error is returned.

Cursors are mostly transactional, i.e. rolling back a (sub-)transaction makes
cursors declared in that subtransaction forgotten in the server.

Committing a transaction or releasing a subtransaction make cursors declared
in this (sub-)transaction or some levels below that propagates the cursor
upwards with only WITH HOLD cursors surviving a committed transaction.

The ECPG runtime library must also follow the server behaviour in tracking
cursors across (sub-)transactions. This is done in misc.c in the ECPGtrans()
function and in cursor.c in the ecpg_commit_cursors() function.
