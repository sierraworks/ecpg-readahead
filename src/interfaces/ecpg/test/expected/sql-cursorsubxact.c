/* Processed by ecpg (regression mode) */
/* These include files are added by the preprocessor */
#include <ecpglib.h>
#include <ecpgerrno.h>
#include <sqlca.h>
/* End of automatic include section */
#define ECPGdebug(X,Y) ECPGdebug((X)+100,(Y))

#line 1 "cursorsubxact.pgc"
#include <stdlib.h>
#include <string.h>


#line 1 "regression.h"






#line 4 "cursorsubxact.pgc"


/* exec sql whenever sqlerror  sqlprint ; */
#line 6 "cursorsubxact.pgc"


#define CURNAME "MyCur"
#define CURNAME_BAD "myCur"

#define QUOTED_CURNAME "\"MyCur\""
#define QUOTED_CURNAME_BAD "\"myCur\""

int main(void)
{
/* exec sql begin declare section */
		
		
		

#line 17 "cursorsubxact.pgc"
 char * curname ;
 
#line 18 "cursorsubxact.pgc"
 int id [ 3 ] ;
 
#line 19 "cursorsubxact.pgc"
  struct varchar_1  { int len; char arr[ 50 ]; }  t [ 3 ] ;
/* exec sql end declare section */
#line 20 "cursorsubxact.pgc"


	ECPGdebug(1, stderr);

	{ ECPGconnect(__LINE__, 0, "regress1" , NULL, NULL , NULL, 0); 
#line 24 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 24 "cursorsubxact.pgc"


	{ ECPGdo(__LINE__, 0, 1, NULL, 0, ECPGst_normal, "create table t1 ( id serial primary key , t text )", ECPGt_EOIT, ECPGt_EORT);
#line 26 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 26 "cursorsubxact.pgc"

	{ ECPGdo(__LINE__, 0, 1, NULL, 0, ECPGst_normal, "insert into t1 ( t ) values ( 'a' ) , ( 'b' ) , ( 'c' )", ECPGt_EOIT, ECPGt_EORT);
#line 27 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 27 "cursorsubxact.pgc"

	{ ECPGtrans(__LINE__, NULL, "commit", 0, 0, 0, NULL);
#line 28 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 28 "cursorsubxact.pgc"


	/*
	 * Cursor name passed in a variable unquoted.
	 * It has to be treated as case-insensitive.
	 */

	printf("test case insensitive cursor name\n\n");

	curname = CURNAME;

	ECPGset_var( 0, &( curname ), __LINE__);\
 /* declare $0 no scroll cursor for select id , t from t1 */
#line 39 "cursorsubxact.pgc"


	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 41 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 41 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case insensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("open %s (case insensitive) succeeded\n", curname);

	curname = CURNAME_BAD;

	{ ECPGclose(__LINE__, 0, 1, NULL, 0, curname, ECPGst_normal, "close $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 49 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 49 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("close %s (case insensitive)  failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("close %s (case insensitive) succeeded\n", curname);

	/*
	 * Cursor name passed in a variable quoted.
	 * It has to be treated as case-sensitive.
	 */

	printf("\ntest case sensitive cursor name\n\n");

	curname = QUOTED_CURNAME;

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 64 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 64 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("open %s (case sensitive) succeeded\n", curname);

	curname = QUOTED_CURNAME_BAD;

	/* Try to close the wrong cursor name, it has to fail with 34000. */
	{ ECPGclose(__LINE__, 0, 1, NULL, 0, curname, ECPGst_normal, "close $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 73 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 73 "cursorsubxact.pgc"
 
	if (sqlca.sqlcode < 0)
		printf("close %s (case sensitive) failed with SQLSTATE %5s (expected 34000)\n", curname, sqlca.sqlstate);
	else
		printf("close %s (case sensitive) succeeded\n", curname);

	/* Try to close the right cursor name, it has to fail with 25P02 */
	curname = QUOTED_CURNAME;
	{ ECPGclose(__LINE__, 0, 1, NULL, 0, curname, ECPGst_normal, "close $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 81 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 81 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("close %s (case sensitive) failed with SQLSTATE %5s (expected 25P02)\n", curname, sqlca.sqlstate);
	else
		printf("close %s (case sensitive) succeeded\n", curname);

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 87 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 87 "cursorsubxact.pgc"


	/*
	 * Try a new transaction and watch the cursor name disappear
	 * after ROLLBACK TO SAVEPOINT.
	 */

	printf("\ntest case sensitive cursor name in savepoint\n\n");

	curname = QUOTED_CURNAME;

	{ ECPGtrans(__LINE__, NULL, "savepoint a", 0, 1, 0, "a");
#line 98 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 98 "cursorsubxact.pgc"


	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 100 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 100 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("close %s (case sensitive) succeeded\n", curname);

	/* The cursor name must be forgotten */
	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 107 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 107 "cursorsubxact.pgc"


	/* It has to fail with 34000 */
	{ ECPGclose(__LINE__, 0, 1, NULL, 0, curname, ECPGst_normal, "close $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 110 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 110 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("close %s failed with SQLSTATE %5s (expected 34000)\n", curname, sqlca.sqlstate);
	else
		printf("close %s succeeded unexpectedly, expected sqlstate 34000\n", curname);

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 116 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 116 "cursorsubxact.pgc"


	/*
	 * Try a new transaction and play games with savepoints.
	 */

	printf("\ntest savepoint handling\n\n");

	{ ECPGtrans(__LINE__, NULL, "savepoint a", 0, 1, 0, "a");
#line 124 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 124 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"savepoint a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"savepoint a\" succeeded\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 130 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 130 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("open %s (case sensitive) succeeded\n", curname);

	{ ECPGtrans(__LINE__, NULL, "savepoint b", 0, 1, 0, "b");
#line 136 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 136 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"savepoint a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"savepoint a\" succeeded\n");

	/* Try to rollback to an unknown savepoint */

	{ ECPGtrans(__LINE__, NULL, "rollback to c", 0, 0, 1, "c");
#line 144 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 144 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to c\" failed with SQLSTATE %5s (expected 3B001)\n", sqlca.sqlstate);
	else
		printf("\"rollback to c\" succeeded unexpectedly (expected 3B001)\n");

	/* "rollback to a" implicitly rolls back "b" */

	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 152 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 152 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"rollback to a\" succeeded\n");

	/* savepoint "b" is now unknown, fails with 3B001 */
	{ ECPGtrans(__LINE__, NULL, "rollback to b", 0, 0, 1, "b");
#line 159 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 159 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to b\" failed with SQLSTATE %5s (expected 3B001)\n", sqlca.sqlstate);
	else
		printf("\"rollback to b\" succeeded unexpectedly (expected 3B001)\n");

	/* "rollback to a" has to succeed again, it's still alive */

	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 167 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 167 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"rollback to a\" succeeded\n");

	/* the cursor was declared under savepoint "a", was forgotten */

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_forward, "1", 0, curname, ECPGst_normal, "fetch from $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, 
	ECPGt_int,&(id[0]),(long)1,(long)1,sizeof(int), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, 
	ECPGt_varchar,&(t[0]),(long)50,(long)1,sizeof(struct varchar_1), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EORT);
#line 175 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 175 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("fetch from %s (case sensitive) failed with SQLSTATE %5s (expected 34000)\n", curname, sqlca.sqlstate);
	else
		printf("fetch from %s (case sensitive) succeeded unexpectedly (expected 34000)\n", curname);

	/* "rollback to a" has to succeed again, it's still alive */

	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 183 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 183 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"rollback to a\" succeeded\n");

	/* "release savepoint a" forgets about the savepoint name */

	{ ECPGtrans(__LINE__, NULL, "release savepoint a", 0, 0, 0, "a");
#line 191 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 191 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"release savepoint a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"release savepoint a\" succeeded\n");

	/* prove the above */

	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 199 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 199 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to a\" failed with SQLSTATE %5s (expected 3B001)\n", sqlca.sqlstate);
	else
		printf("\"rollback to a\" succeeded unexpectedly (expected 3B001)\n");

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 205 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 205 "cursorsubxact.pgc"


	/*
	 * Test the implicit RELEASE SAVEPOINT if a SAVEPOINT
	 * is used with an already existing name.
	 */

	printf("\ntest implicit \"release savepoint a\" for a second \"savepoint a\"\n\n");

	{ ECPGtrans(__LINE__, NULL, "savepoint a", 0, 1, 0, "a");
#line 214 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 214 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"savepoint a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"savepoint a\" succeeded\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 220 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 220 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("open %s (case sensitive) succeeded\n", curname);

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_forward, "1", 0, curname, ECPGst_normal, "fetch from $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, 
	ECPGt_int,&(id[0]),(long)1,(long)1,sizeof(int), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, 
	ECPGt_varchar,&(t[0]),(long)50,(long)1,sizeof(struct varchar_1), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EORT);
#line 226 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 226 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("fetch from %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("fetch from %s (case sensitive) result %d '%s'\n", curname, id[0], t[0].arr);

	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 232 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 232 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"rollback to a\" succeeded\n");

	/* there's no cursor now */

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_forward, "1", 0, curname, ECPGst_normal, "fetch from $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, 
	ECPGt_int,&(id[1]),(long)1,(long)1,sizeof(int), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, 
	ECPGt_varchar,&(t[1]),(long)50,(long)1,sizeof(struct varchar_1), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EORT);
#line 240 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 240 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("fetch from %s (case sensitive) failed with SQLSTATE %5s (expected 34000)\n", curname, sqlca.sqlstate);
	else
		printf("fetch from %s (case sensitive) unexpectedly succeeded (expected 34000)\n", curname);

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 246 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 246 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case sensitive) failed with SQLSTATE %5s (expected 25P02)\n", curname, sqlca.sqlstate);
	else
		printf("open %s (case sensitive) succeeded unexpectedly\n", curname);

	{ ECPGtrans(__LINE__, NULL, "rollback to a", 0, 0, 1, "a");
#line 252 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 252 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"rollback to a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"rollback to a\" succeeded\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 1, 1, 0, curname, ECPGst_normal, "declare $0 no scroll cursor for select id , t from t1", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 258 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 258 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("open %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("open %s (case sensitive) succeeded\n", curname);

	{ ECPGtrans(__LINE__, NULL, "release savepoint a", 0, 0, 0, "a");
#line 264 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 264 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("\"release savepoint a\" failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("\"release savepoint a\" succeeded\n");

	/* release savepoint propagated the cursor to the toplevel transaction */

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_forward, "1", 0, curname, ECPGst_normal, "fetch from $0", 
	ECPGt_char,&(curname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, 
	ECPGt_int,&(id[2]),(long)1,(long)1,sizeof(int), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, 
	ECPGt_varchar,&(t[2]),(long)50,(long)1,sizeof(struct varchar_1), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EORT);
#line 272 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 272 "cursorsubxact.pgc"

	if (sqlca.sqlcode < 0)
		printf("fetch %s (case sensitive) failed with SQLSTATE %5s\n", curname, sqlca.sqlstate);
	else
		printf("fetch %s (case sensitive) result %d '%s'\n", curname, id[2], t[2].arr);

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 278 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 278 "cursorsubxact.pgc"


	/* Drop the test table */

	{ ECPGdo(__LINE__, 0, 1, NULL, 0, ECPGst_normal, "drop table t1", ECPGt_EOIT, ECPGt_EORT);
#line 282 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 282 "cursorsubxact.pgc"

	{ ECPGtrans(__LINE__, NULL, "commit", 0, 0, 0, NULL);
#line 283 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 283 "cursorsubxact.pgc"


	{ ECPGdisconnect(__LINE__, "ALL");
#line 285 "cursorsubxact.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 285 "cursorsubxact.pgc"


	return 0;
}
