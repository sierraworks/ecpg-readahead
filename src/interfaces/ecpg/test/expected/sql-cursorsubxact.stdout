test case insensitive cursor name

open MyCur (case insensitive) succeeded
close myCur (case insensitive) succeeded

test case sensitive cursor name

open "MyCur" (case sensitive) succeeded
close "myCur" (case sensitive) failed with SQLSTATE 34000 (expected 34000)
close "MyCur" (case sensitive) failed with SQLSTATE 25P02 (expected 25P02)

test case sensitive cursor name in savepoint

close "MyCur" (case sensitive) succeeded
close "MyCur" failed with SQLSTATE 34000 (expected 34000)

test savepoint handling

"savepoint a" succeeded
open "MyCur" (case sensitive) succeeded
"savepoint a" succeeded
"rollback to c" failed with SQLSTATE 3B001 (expected 3B001)
"rollback to a" succeeded
"rollback to b" failed with SQLSTATE 3B001 (expected 3B001)
"rollback to a" succeeded
fetch from "MyCur" (case sensitive) failed with SQLSTATE 34000 (expected 34000)
"rollback to a" succeeded
"release savepoint a" succeeded
"rollback to a" failed with SQLSTATE 3B001 (expected 3B001)

test implicit "release savepoint a" for a second "savepoint a"

"savepoint a" succeeded
open "MyCur" (case sensitive) succeeded
fetch from "MyCur" (case sensitive) result 1 'a'
"rollback to a" succeeded
fetch from "MyCur" (case sensitive) failed with SQLSTATE 34000 (expected 34000)
open "MyCur" (case sensitive) failed with SQLSTATE 25P02 (expected 25P02)
"rollback to a" succeeded
open "MyCur" (case sensitive) succeeded
"release savepoint a" succeeded
fetch "MyCur" (case sensitive) result 1 'a'
